### ISSUE TITLE
[question] `CONAN_V2_MODE` and `settings.os` access in `source()` method

### ISSUE THREAD

**Issue by MarcelRaad** [conan-io/docs#2458]

**Comment by spectras** [comment]
Since this issue was not closed, I take the opportunity for this comment:

> Otherwise, it is recommended to get the sources in the build() method.

The recipe tools `patch` and `apply_conandata_patches` from `conan.tools.files` are currently [hardcoded](https://github.com/conan-io/conan/blob/c1557344eb4f6918129034091ecfce7971e9fd13/conan/tools/files/patches.py#L61) to patch in the `source_folder`.

So if one has different sources depending on the platform, and needs to patch them, they are out of luck.

**Comment by memsharded** [comment]
But @spectras the ``apply_conandata_patches()`` already assumes a structure in the ``conandata.yml`` of the patches that is platform independent, so it should be able to be applied in the ``source()`` method. In other words, ``apply_conandata_patches()`` always applies the same patches, so it can be applied in the ``source()`` safely, or otherwise, a custom solution with conditionals is necessary (and that is more difficult to define a common structure, because patches can vary per OS, per compiler, etc)

**Comment by spectras** [comment]
That is true, it was more a nice to have, to be able to customize the behavior a bit and avoid writing a custom patching logic for simple cases. But this does raise the question of when it stops being a simple case.<br>
Forget it, we had the case (as we are moving download of proprietary things from `source()` to `build()`, preparing for conan2), and we thought it could be a nice enhancement. But creating complex configuration on conan side is probably not worth the effort indeed. Thanks for the fast answer!
